{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "041a4142-45b0-4fc1-90c4-9897851feb5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# seq_orf.ipynb\n",
    "\n",
    "# Cell 1\n",
    "\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "\n",
    "AMINO_ACIDS = {\n",
    "    (\"Ala\", \"A\"): [\"GCT\", \"GCA\", \"GCC\", \"GCG\"],  # Alanine\n",
    "    (\"Arg\", \"R\"): [\"CGT\", \"CGC\", \"CGA\", \"CGG\", \"AGA\", \"AGG\"],  # Arginine\n",
    "    (\"Asn\", \"N\"): [\"AAT\", \"AAC\"],  # Asparagine\n",
    "    (\"Asp\", \"D\"): [\"GAT\", \"GAC\"],  # Aspartic Acid\n",
    "    (\"Cys\", \"C\"): [\"TGT\", \"TGC\"],  # Cysteine\n",
    "    (\"Gln\", \"Q\"): [\"CAA\", \"CAG\"],  # Glutamine\n",
    "    (\"Glu\", \"E\"): [\"GAA\", \"GAG\"],  # Glutamic Acid\n",
    "    (\"Gly\", \"G\"): [\"GGT\", \"GGC\", \"GGA\", \"GGG\"],  # Glycine\n",
    "    (\"His\", \"H\"): [\"CAT\", \"CAC\"],  # Histidine\n",
    "    (\"Ile\", \"I\"): [\"ATT\", \"ATC\", \"ATA\"],  # Isoleucine\n",
    "    (\"Leu\", \"L\"): [\"TTA\", \"TTG\", \"CTT\", \"CTC\", \"CTA\", \"CTG\"],  # Leucine\n",
    "    (\"Lys\", \"K\"): [\"AAA\", \"AAG\"],  # Lysine\n",
    "    (\"Met\", \"M\"): [\"ATG\"],  # Methionine (Start)\n",
    "    (\"Phe\", \"F\"): [\"TTT\", \"TTC\"],  # Phenylalanine\n",
    "    (\"Pro\", \"P\"): [\"CCT\", \"CCC\", \"CCA\", \"CCG\"],  # Proline\n",
    "    (\"Ser\", \"S\"): [\"TCT\", \"TCC\", \"TCA\", \"TCG\", \"AGT\", \"AGC\"],  # Serine\n",
    "    (\"Thr\", \"T\"): [\"ACT\", \"ACC\", \"ACA\", \"ACG\"],  # Threonine\n",
    "    (\"Trp\", \"W\"): [\"TGG\"],  # Tryptophan\n",
    "    (\"Tyr\", \"Y\"): [\"TAT\", \"TAC\"],  # Tyrosine\n",
    "    (\"Val\", \"V\"): [\"GTT\", \"GTC\", \"GTA\", \"GTG\"],  # Valine\n",
    "    (\"Stop\", \"0\"): [\"TAA\", \"TGA\", \"TAG\"],  # Stop\n",
    "}\n",
    "\n",
    "\n",
    "def reverse_complement(seq):\n",
    "    seq_reverse = \"\"\n",
    "    for c in seq[::-1]:\n",
    "        if c == \"C\":\n",
    "            seq_reverse += \"G\"\n",
    "        elif c == \"G\":\n",
    "            seq_reverse += \"C\"\n",
    "        elif c == \"A\":\n",
    "            seq_reverse += \"T\"\n",
    "        elif c == \"T\":\n",
    "            seq_reverse += \"A\"\n",
    "    return seq_reverse\n",
    "\n",
    "\n",
    "def make_codons(seq, offset):\n",
    "    codons = []\n",
    "    # Each codon is a grouping of three successive nucleotides,\n",
    "    # starting at some offset in a DNA/RNA sequence\n",
    "    # Each codon encodes a specific amino acid\n",
    "    for i in range(offset, len(seq), 3):\n",
    "        if i + 3 <= len(seq):\n",
    "            codons.append(seq[i : i + 3])\n",
    "    return codons\n",
    "\n",
    "\n",
    "def find_codon(codon_list, codon):\n",
    "    try:\n",
    "        # Get index if this codon appears in the codon list\n",
    "        idx = codon_list.index(codon)\n",
    "        return idx\n",
    "    except ValueError:\n",
    "        # This codon does not appear in the list\n",
    "        return -1\n",
    "\n",
    "\n",
    "def decode_codons(codon_string):\n",
    "    # Invert AMINO_ACIDS so it is keyed by codon, not amino acid\n",
    "    inverted_dict = {}\n",
    "    for k in AMINO_ACIDS:\n",
    "        for v in AMINO_ACIDS[k]:\n",
    "            inverted_dict[v] = k\n",
    "    # Build string of single-letter amino acids based upon each codon\n",
    "    acids = \"\"\n",
    "    for c in codon_string.split():\n",
    "        # The single letter is the 2nd element in the tuple: [1]\n",
    "        acids += inverted_dict[c][1]\n",
    "    # Don't include start and stop codons in amino acid sequence\n",
    "    acids = acids[1:-1]\n",
    "    return acids\n",
    "\n",
    "\n",
    "def get_orf(codon_string, offset):\n",
    "    # Split sequence into codon list starting at offset\n",
    "    codon_list = make_codons(codon_string, offset)\n",
    "\n",
    "    # Find possible index for START codon\n",
    "    start_idx = find_codon(codon_list, \"ATG\")\n",
    "    if start_idx < 0:\n",
    "        return None\n",
    "\n",
    "    # Find possible indexes for all three STOP codons\n",
    "    stop_idexes = [\n",
    "        find_codon(codon_list, \"TAA\"),\n",
    "        find_codon(codon_list, \"TAG\"),\n",
    "        find_codon(codon_list, \"TGA\"),\n",
    "    ]\n",
    "\n",
    "    # Remove any STOP codon index that comes before the START index\n",
    "    stop_idexes[:] = [idx for idx in stop_idexes if idx > start_idx]\n",
    "    if len(stop_idexes) == 0:\n",
    "        return None\n",
    "\n",
    "    # Use the index of the first occuring valid STOP codon\n",
    "    stop_idx = min(stop_idexes)\n",
    "\n",
    "    # An empty frame if STOP immediately follows START\n",
    "    if stop_idx == start_idx + 1:\n",
    "        return None\n",
    "\n",
    "    # Build a string of all codons (including START and STOP)\n",
    "    codon_string = \"\"\n",
    "    for idx in range(start_idx, stop_idx + 1):\n",
    "        codon_string += codon_list[idx] + \" \"\n",
    "\n",
    "    # If an ORF exists, also display codons as single character amino acids\n",
    "    if len(codon_string) > 0:\n",
    "        codon_string += f\" ({decode_codons(codon_string)})\"\n",
    "\n",
    "    return codon_string\n",
    "\n",
    "\n",
    "def analyze_file(file_name):\n",
    "    with open(file_name, \"rb\") as f_in:\n",
    "        print(f\"Analyzing {file_name} . . .\")\n",
    "\n",
    "        # Read in text file into an array of file bytes\n",
    "        f_bytes = bytearray(f_in.read())\n",
    "\n",
    "        # Enforce uppercase and remove non-letters, convert to UTF-8\n",
    "        seq = bytearray(f_bytes).decode().upper()\n",
    "        seq = re.compile(\"[^A-Z]\").sub(\"\", seq)\n",
    "\n",
    "        # Print the original given sequence\n",
    "        print(f\"Original sequence:  {seq} \")\n",
    "\n",
    "        # Build and print the reverse compliment of the given sequence\n",
    "        seq_rc = reverse_complement(seq)\n",
    "        print(f\"Reverse complement: {seq_rc} \")\n",
    "\n",
    "        # Print any open reading frames in forward sequence\n",
    "        if s := get_orf(seq, 0):\n",
    "            print(f\"Open Frame +1: {s}\")\n",
    "        if s := get_orf(seq, 1):\n",
    "            print(f\"Open Frame +2: {s}\")\n",
    "        if s := get_orf(seq, 2):\n",
    "            print(f\"Open Frame +3: {s}\")\n",
    "\n",
    "        # Print any open reading frames in reverse compliment of sequence\n",
    "        if s := get_orf(seq_rc, 0):\n",
    "            print(f\"Open Frame -1: {s}\")\n",
    "        if s := get_orf(seq_rc, 1):\n",
    "            print(f\"Open Frame -2: {s}\")\n",
    "        if s := get_orf(seq_rc, 2):\n",
    "            print(f\"Open Frame -3: {s}\")\n",
    "\n",
    "        return\n",
    "\n",
    "\n",
    "analyze_file(\"seq2.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce74f6bb-4ec3-4cbc-b53d-3afa187ed78d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell 2 - Another example sequence\n",
    "\n",
    "analyze_file(\"seq3.txt\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
